### DISCUSSION TITLE
Should critical adviser messages drop the whole queue?

### DISCUSSION THREAD

**Discussion by lewri** [CorsixTH/CorsixTH#2425] [Research Room]
Currently when we give the advisor a critical message using the `override_current` flag we drop the entire adviser queue as part of the process:
https://github.com/CorsixTH/CorsixTH/blob/87858c2d5a4060a6039ca4110fb24403d17d4123/CorsixTH/Lua/dialogs/adviser.lua#L156-L174

It may be that this behaviour is a bit overkill for what is actually useful to the player. Sure, we want that message played now, but it may also be that the other adviser messages are still useful that were in the queue.

Discussion here is if the adviser should instead play the critical message immediately but retain the existing queue.
Another hole I just discussed in that multi-if statement is that the `override_current` flag can be thrown away if the adviser is up and not talking -- the message with the flag will be queued normally at the end of the message queue?

**Comment by Alberth289346** [comment]
> Discussion here is if the adviser should instead play the critical message immediately but retain the existing queue.

I would say dropping the entire queue is way too much power at the calling side. Code doesn't track what messages are currently queued, so the "override" is just blindly discarding messages without even knowing if that is a good idea.  
For example, there may an earthquake happening, an inspector for the epidemic may be arriving, and the operating room may be flooded with a patient on the table :) Is it a good idea to just  delete all those messages in favor of some new message?

As far as I can see, at the advisor is the only spot where you can sanely decide about order and/or discarding of messages. The caller can set the priority of the message and that should be sufficient imho.  
The code already seems to sort on priority (probably?) for the next message. As such, I don't understand why other messages need to be deleted at all. The only other possible meaning I think is to literally abort the current message, but then one should at least check that the current message  is not also a critical message is my guess.

**Reply by ageekhere** [reply]
Is there a adviser message history that the player can view? I have seen messages get closed for another message before i had a chance to read it.

**Reply by MandrakeHorse** [reply]
That would be a neat enhancement. Perhaps there could be an option in the Charts dropdown (found in the menu bar at the top) which opens a popup window containing the Adviser's most recent messages (20 or 50 of them?) and a scrollbar. The main challenges are that it would probably need to be built from scratch since I don't think Theme Hospital's UI has anything for text logs, and we'd need to make sure the Adviser's messages are being temporarily logged.

**Reply by Alberth289346** [reply]
OpenTTD also has a pop-up with messages. There you can click at the bar below it, and the previous message appears (and if you click again the next previous, etc). I don't know if the original game also had that.

That could be another solution here.

**Reply by MandrakeHorse** [reply]
A method of reviewing the Adviser's recent messages has already been proposed as an enhancement, with the suggestion that the Dynamic Info Bar could be repurposed for this task: https://github.com/CorsixTH/CorsixTH/issues/530

**Reply by Alberth289346** [reply]
Ah great, a previous discussion :)
